std/mutex.zig

```diff
@@ -3,16 +3,31 @@ const builtin = @import("builtin");
 const AtomicOrder = builtin.AtomicOrder;
 const AtomicRmwOp = builtin.AtomicRmwOp;
 const assert = std.debug.assert;
+const linux = std.os.linux;
+
+// Reading: Futexes Are Tricky by Ulrich Drepper https://www.akkadia.org/drepper/futex.pdf
+
+// TODO robust mutexes https://www.kernel.org/doc/Documentation/robust-futexes.txt
 
-/// TODO use syscalls instead of a spinlock
 pub const Mutex = struct {
-    lock: u8, // TODO use a bool
+    // TODO: Windows and OSX with futex equivilents
+    // 0: unlocked
+    // 1: locked, no waiters
+    // 2: locked: one or more waiters
+    lock: u32, // futexs are 32-bits on all architectures
 
     pub const Held = struct {
         mutex: *Mutex,
 
         pub fn release(self: Held) void {
-            assert(@atomicRmw(u8, &self.mutex.lock, builtin.AtomicRmwOp.Xchg, 0, AtomicOrder.SeqCst) == 1);
+            if (@atomicRmw(u32, &self.mutex.lock, AtomicRmwOp.Sub, 1, AtomicOrder.Release) != 1) {
+                self.mutex.lock = 0;
```

Comment:
You can't have a naked write racing with an atomic write (the cmpxchg on line 42).

Reply:
Why doesn't the `@atomicRmw` do an xchg and set the value to 0? You can still check if the previous value was 1.